In [8]:
import pandas as pd
import sqlalchemy
import polars as pl 

engine = sqlalchemy.create_engine('postgresql://verzochia:verzochia@localhost:5432/db_v1')

pd.set_option('display.max_columns', 100)


In [9]:

fchap = pd.read_parquet('./parquet/Fonction_Chapitre_2020.parquet')
fcompte = pd.read_parquet('./parquet/Fonction_Compte_2020.parquet')
fref = pd.read_parquet('./parquet/Fonction_Referentiel_2020.parquet')
df_table_budget = pd.read_sql('Select * from bloc_budget limit 1500000', engine)

------------ 
Données lignes budget 

----------------

In [10]:
df_table_budget['Id_Fichier'].drop_duplicates().shape[0]

11239

In [12]:
df_table_budget.columns

Index(['id', 'Id_Fichier', 'Nomenclature', 'Exer', 'TypOpBudg', 'Operation',
       'Nature', 'ContNat', 'LibCpte', 'Fonction', 'ContFon', 'ArtSpe',
       'CodRD', 'MtBudgPrec', 'MtRARPrec', 'MtPropNouv', 'MtPrev', 'OpBudg',
       'CredOuv', 'MtReal', 'MtRAR3112', 'ContOp', 'OpeCpteTiers', 'MtSup',
       'MtSup_APVote', 'MtSup_Brut', 'MtSup_BudgetHorsRAR', 'MtSup_Comp',
       'MtSup_ICNE', 'MtSup_ICNEPrec', 'MtSup_MtOpeCumul', 'MtSup_MtOpeInfo',
       'MtSup_Net', 'MtSup_ProdChaRat', 'MtSup_RARPrec', 'CaracSup',
       'CaracSup_TypOpe', 'CaracSup_Section', 'CaracSup_ChapSpe',
       'CaracSup_ProgAutoLib', 'CaracSup_ProgAutoNum', 'CaracSup_VirCredNum',
       'CaracSup_CodeRegion'],
      dtype='object')

In [13]:
df_table_budget.dtypes

id                        int64
Id_Fichier               object
Nomenclature             object
Exer                      int64
TypOpBudg                object
Operation                object
Nature                   object
ContNat                  object
LibCpte                  object
Fonction                 object
ContFon                  object
ArtSpe                   object
CodRD                    object
MtBudgPrec              float64
MtRARPrec               float64
MtPropNouv              float64
MtPrev                  float64
OpBudg                   object
CredOuv                 float64
MtReal                  float64
MtRAR3112               float64
ContOp                   object
OpeCpteTiers             object
MtSup                    object
MtSup_APVote            float64
MtSup_Brut              float64
MtSup_BudgetHorsRAR     float64
MtSup_Comp               object
MtSup_ICNE              float64
MtSup_ICNEPrec          float64
MtSup_MtOpeCumul        float64
MtSup_Mt

In [14]:
df_budget_fonction = df_table_budget.dropna(subset=['Fonction'])
df_9 = df_budget_fonction[df_budget_fonction['Fonction'].str.startswith('9')]
df_9[['Id_Fichier', 'Nomenclature', 'Exer', 'Fonction', 'ContFon']]

Id_Fichier         Nomenclature  Exer Fonction ContFon
282         837742  M14-M14_COM_SUP3500  2022       90    None
317         837742  M14-M14_COM_SUP3500  2022        9    None
364         837742  M14-M14_COM_SUP3500  2022       90    None
449         837742  M14-M14_COM_SUP3500  2022       90    None
459         837742  M14-M14_COM_SUP3500  2022       90    None
...            ...                  ...   ...      ...     ...
1498844     648309  M14-M14_COM_SUP3500  2020       91    None
1498953     648309  M14-M14_COM_SUP3500  2020       90    None
1499447     648309  M14-M14_COM_SUP3500  2020       94    None
1499470     648309  M14-M14_COM_SUP3500  2020       94    None
1499521     648309  M14-M14_COM_SUP3500  2020       94    None

[61272 rows x 5 columns]

Sur les X Lignes budget, parmis les fonctions commencant par 9, plusieurs fonctions n'ont pas de correspondance dans les libelles Fonction_Compte 

In [25]:
df_9_compte = df_9.merge(fcompte, left_on= ['Nomenclature', 'Fonction'], right_on= ['Nomenclature', 'Code'], how = 'left')
df_9_compte1 = df_9_compte[['Id_Fichier', 'Exer_x', 'Exer_y','Nomenclature', 'Nature', 'Fonction', 'Code', 'Libelle']]
df_9_na = df_9_compte[df_9_compte['Libelle'].isna()].rename(columns = {'Code' : 'Fc_Code', 'Libelle' : 'Fc_Libelle'})
liste_code_9_na = df_9_na['Fonction'].drop_duplicates().to_list()
liste_code_9_na

['9',
 '94',
 '96',
 '93201',
 '93501',
 '91',
 '97',
 '928',
 '9591',
 '952',
 '951',
 '959',
 '961',
 '90',
 '9001',
 '941',
 '98']

In [26]:
df_table_budget[df_table_budget['Fonction'] == '93501']

id Id_Fichier Nomenclature  Exer TypOpBudg Operation Nature  \
68069  1914380     805749      M57-M57  2022      None      None   6234   

      ContNat LibCpte Fonction ContFon ArtSpe    CodRD  MtBudgPrec  MtRARPrec  \
68069    None    None    93501     935  False  Dépense     24834.0        0.0   

       MtPropNouv  MtPrev                     OpBudg  CredOuv  MtReal  \
68069         0.0     0.0  Opération réelle ou mixte  24834.0     0.0   

       MtRAR3112 ContOp OpeCpteTiers MtSup  MtSup_APVote  MtSup_Brut  \
68069        0.0   None         None  None           NaN         NaN   

       MtSup_BudgetHorsRAR MtSup_Comp  MtSup_ICNE  MtSup_ICNEPrec  \
68069                  NaN       None         NaN             NaN   

       MtSup_MtOpeCumul  MtSup_MtOpeInfo  MtSup_Net MtSup_ProdChaRat  \
68069               NaN              NaN        NaN             None   

      MtSup_RARPrec                                           CaracSup  \
68069          None  [{"@Code": "ProgAutoNum", "@V": "AE"}, {"@Code...   

       CaracSup_TypOpe CaracSup_Section CaracSup_ChapSpe CaracSup_ProgAutoLib  \
68069              NaN             None             None                LibAE   

      CaracSup_ProgAutoNum CaracSup_VirCredNum  CaracSup_CodeRegion  
68069                   AE                None                  NaN

In [46]:
liste_vide_pur = []
for i in liste_code_9_na : 
  type_compte = fcompte[fcompte['Code'] == i]
  type_chap = fchap[fchap['Code'] == i]
  type_ref = fref[fref['Code'] == i]
  if type_compte.shape[0] > 0 : 
    print(i, 'present dans fcompte')
  elif type_chap.shape[0] > 0 : 
    print(i, 'present dans fchap')
  elif type_ref.shape[0] > 0 : 
    print(i, 'present dans referentiel')
  else : 
    liste_vide_pur.append(i)

9 present dans referentiel
94 present dans fcompte
96 present dans referentiel
91 present dans fcompte
928 present dans fcompte
952 present dans fcompte
951 present dans fcompte
90 present dans fcompte
9001 present dans fcompte
941 present dans fcompte


In [47]:
liste_vide_pur

['93201', '93501', '97', '9591', '959', '961', '98']

In [71]:
df_table_budget[df_table_budget['Fonction'] == '961'][['Nomenclature', 'Exer']]

Nomenclature  Exer
876818  M14-M14_COM_SUP3500  2021
876906  M14-M14_COM_SUP3500  2021
876950  M14-M14_COM_SUP3500  2021
877037  M14-M14_COM_SUP3500  2021
877327  M14-M14_COM_SUP3500  2021

N'existent pas : 

- 9591 => 591 n'existe pas et 91 a deux val (compte et ref)
- 959 => 59 Provisions pour dépréciation des comptes financiers
- 961 ==> 61 a deux val différentes. Services extérieurs (compte) et Services en faveur des personnes âgées (ref)
- 98 

In [73]:
#generalisons 
df_t_compte = df_table_budget.merge(fcompte, left_on= ['Nomenclature', 'Fonction'], right_on= ['Nomenclature', 'Code'], how = 'left')
df_t_compte = df_t_compte[['Id_Fichier', 'Exer_x', 'Exer_y','Nomenclature', 'Nature', 'Fonction', 'Code', 'Libelle']]
df_t_compte_na = df_t_compte[df_t_compte['Libelle'].isna()].rename(columns = {'Code' : 'Fc_Code', 'Libelle' : 'Fc_Libelle'})
liste_t_compte_na = df_t_compte_na['Fonction'].drop_duplicates().to_list()
liste_t_compte_na


[None,
 '020',
 '821',
 '520',
 '01',
 '414',
 '524',
 '321',
 '810',
 '411',
 '71',
 '112',
 '511',
 '812',
 '9',
 '413',
 '40',
 '023',
 '20',
 '824',
 '412',
 '814',
 '823',
 '510',
 '113',
 '833',
 '822',
 '422',
 '211',
 '251',
 '212',
 '314',
 '64',
 '813',
 '423',
 '024',
 '60',
 '811',
 '025',
 '253',
 '421',
 '021',
 '026',
 '213',
 '2',
 '33',
 '94',
 '831',
 '61',
 '288',
 '323',
 '11',
 '4228',
 '632',
 '281',
 '311',
 '18',
 '313',
 '70',
 '428',
 '10',
 '048',
 '312',
 '410',
 '4221',
 '420',
 '022',
 '30',
 '338',
 '315',
 '4238',
 '316',
 '50',
 '031',
 '4222',
 '331',
 '283',
 '588',
 '424',
 '325',
 '326',
 '317',
 '518',
 '844',
 '4212',
 '52',
 '02',
 '830',
 '816',
 '110',
 '820',
 '324',
 '415',
 '523',
 '512',
 '521',
 '255',
 '815',
 '12',
 '63',
 '254',
 '22',
 '23',
 '24',
 '832',
 '522',
 '54',
 '76',
 '843',
 '4213',
 '447',
 '221',
 '430',
 '80',
 '446',
 '425',
 '444',
 '731',
 '57',
 '555',
 '828',
 '6311',
 '4211',
 '633',
 '62',
 '442',
 '051',
 '842',


In [75]:
liste_vide_total = []
for i in liste_t_compte_na : 
  type_compte = fcompte[fcompte['Code'] == i]
  type_chap = fchap[fchap['Code'] == i]
  type_ref = fref[fref['Code'] == i]
  if type_compte.shape[0] > 0 : 
    pass 
  elif type_chap.shape[0] > 0 : 
    pass
  elif type_ref.shape[0] > 0 : 
    pass
  else : 
    liste_vide_total.append(i)

In [81]:
len(liste_vide_total)

173

On a des cas similaire pour les natures, pas que dans les 9 : cas nature == 281532

In [89]:
df_table_budget[(df_table_budget['Fonction'].isin(liste_vide_total)) & 
        (df_table_budget['Exer'] == 2020) & 
        (~df_table_budget['Fonction'].isna())].drop_duplicates(['Nomenclature', 'Exer', 'Fonction'])

id Id_Fichier         Nomenclature  Exer TypOpBudg Operation  \
186380   2032654     690975  M14-M14_COM_SUP3500  2020      None      9019   
186383   2032657     690975  M14-M14_COM_SUP3500  2020      None      9019   
214820   2061130     641951  M14-M14_COM_SUP3500  2020      None       901   
214829   2061139     641951  M14-M14_COM_SUP3500  2020      None       911   
214830   2061140     641951  M14-M14_COM_SUP3500  2020      None       911   
...          ...        ...                  ...   ...       ...       ...   
1461812   796058     732400  M14-M14_COM_SUP3500  2020      None       806   
1461814   796060     732400  M14-M14_COM_SUP3500  2020      None       817   
1461822   796068     732400  M14-M14_COM_SUP3500  2020      None       805   
1461825   796071     732400  M14-M14_COM_SUP3500  2020      None       800   
1461878   796126     732400  M14-M14_COM_SUP3500  2020      None      None   

        Nature ContNat                                   LibCpte Fonction  \
186380    2051    9019                         MATERIEL SCOLAIRE     2121   
186383    2183    9019                         MATERIEL SCOLAIRE     2122   
214820     202     901                          ETUDES URBANISME    02011   
214829     205     911  ACQUISITION MATERIEL OUTILLAGE ET DIVERS    02021   
214830     205     911  ACQUISITION MATERIEL OUTILLAGE ET DIVERS     4148   
...        ...     ...                                       ...      ...   
1461812   2031    None                       Aires de compostage    81231   
1461814   2031    None                         RESEAU DE CHALEUR     8121   
1461822   2051    None               QUAI DE TRANSFERT DE LESCAR     8126   
1461825   2145    None    Aménagement Centre de Tri de Sévignacq    81221   
1461878  60632    None                                      None     8125   

        ContFon ArtSpe    CodRD  MtBudgPrec  MtRARPrec  MtPropNouv  MtPrev  \
186380     None  False  Dépense         NaN        NaN         NaN     0.0   
186383     None  False  Dépense         NaN        NaN         NaN     0.0   
214820     None  False  Dépense         NaN        NaN         NaN     0.0   
214829     None  False  Dépense         NaN        NaN         NaN     0.0   
214830     None  False  Dépense         NaN        NaN         NaN     0.0   
...         ...    ...      ...         ...        ...         ...     ...   
1461812    None  False  Dépense         NaN        NaN         NaN     NaN   
1461814    None  False  Dépense         NaN        NaN         NaN     NaN   
1461822    None  False  Dépense         NaN        NaN         NaN     NaN   
1461825    None  False  Dépense         NaN        NaN         NaN     NaN   
1461878    None  False  Dépense         NaN        NaN         NaN     NaN   

                            OpBudg   CredOuv    MtReal  MtRAR3112 ContOp  \
186380   Opération réelle ou mixte       NaN       NaN        NaN   None   
186383   Opération réelle ou mixte       NaN       NaN        NaN   None   
214820   Opération réelle ou mixte       NaN       NaN        NaN   None   
214829   Opération réelle ou mixte       NaN       NaN        NaN   None   
214830   Opération réelle ou mixte       NaN       NaN        NaN   None   
...                            ...       ...       ...        ...    ...   
1461812  Opération réelle ou mixte   23075.0  23075.00        0.0   None   
1461814  Opération réelle ou mixte       NaN       NaN        0.0   None   
1461822  Opération réelle ou mixte      30.0    670.00        0.0   None   
1461825  Opération réelle ou mixte       NaN  17074.20        0.0   None   
1461878  Opération réelle ou mixte  146400.0  25731.96        0.0   None   

        OpeCpteTiers                                      MtSup  MtSup_APVote  \
186380          None   {"@Code": "MtOpeCumul", "@V": "1031.00"}           NaN   
186383          None  {"@Code": "MtOpeCumul", "@V": "17520.31"}           NaN   
214820          None  {"@Code": "MtOpeCumul", "@V": "91612.

Par contre, Fonction = 2121 existe pas dans les df mais dans sur le site, mais pas 02011 

APr_s verif du fichier scrapé.. bah il est dans le xml.... l'extraction est un échec. 

In [54]:
print(fchap[fchap['Code'].str.contains('97')].shape[0])
print(fcompte[fcompte['Code'].str.contains('97')].shape[0])
print(fref[fref['Code'].str.contains('97')].shape[0])

0
0
0


In [ ]:
df_table_budget[df_table_budget['Fonction'] == '97'][['Nomenclature', 'Exer']]

Beaucoup de mal à croire que 97 n'existe pas 

Non mais quel con, j'ai scrapé 2020 et pas 2021, 97 correspond au plan de relance crise sanitaire de fref 

In [63]:
df_table_budget[df_table_budget['Fonction'] == '9591'][['Nomenclature', 'Exer']]

Nomenclature  Exer
68069      M57-M57  2022

Par contre 9591 n'existe pas dans les fichiers de 2021 / 2020

In [62]:
df_table_budget[df_table_budget['Fonction'] == '93201'][['Nomenclature', 'Exer']]

Nomenclature  Exer
67932      M57-M57  2022
68061      M57-M57  2022
68062      M57-M57  2022

Existe en 2022

In [16]:
df_9_na[['Id_Fichier', 'Nomenclature', 'Exer_x', 'Fonction', 'ContFon', 'Fc_Code']].head(10)

Id_Fichier         Nomenclature  Exer_x Fonction ContFon Fc_Code
1       837742  M14-M14_COM_SUP3500    2022        9    None     NaN
8       837742  M14-M14_COM_SUP3500    2022        9    None     NaN
10      837742  M14-M14_COM_SUP3500    2022        9    None     NaN
16      653716  M14-M14_COM_SUP3500    2020       94    None     NaN
21      653716  M14-M14_COM_SUP3500    2020       94    None     NaN
136     763372  M14-M14_COM_SUP3500    2020       94    None     NaN
139     763372  M14-M14_COM_SUP3500    2020       94    None     NaN
141     763372  M14-M14_COM_SUP3500    2020       94    None     NaN
142     763372  M14-M14_COM_SUP3500    2020       94    None     NaN
150     763372  M14-M14_COM_SUP3500    2020       94    None     NaN

In [20]:
fcompte[fcompte['Code'] == '94']

Exer Nomenclature Code DEquip  DOES  DOIS    DR  \
333  2020      M52-M52   94   None  None  None  None   
477  2020      M57-M57   94      0                     
231  2020      M71-M71   94   None  None  None  None   

                         Lib_court                        Libelle REquip  \
333  Services communs non ventilés  Services communs non ventilés   None   
477  Services communs non ventilés  Services communs non ventilés      0   
231  Services communs non ventilés  Services communs non ventilés   None   

     ROES  ROIS    RR RegrTotalise Supprime SupprimeDepuis  
333  None  None  None            1     None           None  
477                              1     None           None  
231  None  None  None            1     None           None

Est ce que ces codes sont dans libelle fonction chapitre ? 

In [22]:
fchap[fchap['Code']== '94']

Empty DataFrame
Columns: [Exer, Nomenclature, Code, Lib_court, Libelle, Section, Special, TypeChapitre]
Index: []

In [24]:
fref[fref['Code'] == '94']

Exer         Nomenclature Code                                 Lib_court  \
109  2020              M52-M52   94                 Développement touristique   
110  2020              M71-M71   94  Industrie, artisanat, commerce et autres   
109  2020  M14-M14_COM_SUP3500   94      Aides commerce et services marchands   

                                               Libelle  
109                          Développement touristique  
110  Industrie, artisanat, commerce et autres services  
109        Aides au commerce et aux services marchands

Peux on simplement concat compte et ref ? 

In [211]:
compte_ref = pd.concat([fcompte, fref])
compte_ref.duplicated(['Nomenclature', 'Code']).sum()

17

In [218]:
compte_ref[compte_ref.duplicated(['Nomenclature', 'Code'], keep = False)].sort_values(['Nomenclature', 'Code']).head(10)

Exer         Nomenclature Code DEquip  DOES  DOIS    DR  \
0    2020  M14-M14_COM_SUP3500   90   None  None  None  None   
105  2020  M14-M14_COM_SUP3500   90    NaN   NaN   NaN   NaN   
112  2020  M14-M14_COM_SUP3500   91   None  None  None  None   
106  2020  M14-M14_COM_SUP3500   91    NaN   NaN   NaN   NaN   
123  2020  M14-M14_COM_SUP3500   92   None  None  None  None   
107  2020  M14-M14_COM_SUP3500   92    NaN   NaN   NaN   NaN   
235  2020  M14-M14_COM_SUP3500   93   None  None  None  None   
108  2020  M14-M14_COM_SUP3500   93    NaN   NaN   NaN   NaN   
244  2020  M14-M14_COM_SUP3500   95      0                     
110  2020  M14-M14_COM_SUP3500   95    NaN   NaN   NaN   NaN   

                                    Lib_court  \
0                     Opérations d'équipement   
105                 Interventions économiques   
112                  Opérations non ventilées   
106                         Foires et marchés   
123                   Services individualisés   
107   Aides à l'agriculture et aux industries   
235             Services communs non ventilés   
108  Aides à l'énergie, indus. manufact., BTP   
244   Produits des cessions d'immobilisations   
110                         Aides au tourisme   

                                               Libelle REquip  ROES  ROIS  \
0                              Opérations d'équipement   None  None  None   
105                          Interventions économiques    NaN   NaN   NaN   
112                           Opérations non ventilées   None  None  None   
106                                  Foires et marchés    NaN   NaN   NaN   
123                            Services individualisés   None  None  None   
107  Aides à l'agriculture et aux industries agro-a...    NaN   NaN   NaN   
235                      Services communs non ventilés   None  None  None   
108  Aides à l'énergie, aux industries manufacturiè...    NaN   NaN   NaN   
244            Produits des cessions d'immobilisations      0               
110                                  Aides au tourisme    NaN   NaN   NaN   

       RR RegrTotalise Supprime SupprimeDepuis  
0    None            1     None           None  
105   NaN          NaN      NaN            NaN  
112  None         None     None           None  
106   NaN          NaN      NaN            NaN  
123  None            1     None           None  
107   NaN          NaN      NaN            NaN  
235  None            1     None           None  
108   NaN          NaN      NaN            NaN  
244    95         None     None           None  
110   NaN          NaN      NaN            NaN

Malheureusement, ces codes qui se chevauchent possèdent des libelles différents. On ne peut pas se baser sur contfon pour les séparer, à première vue on ne peut pas se baser sur une autre colonne non plus 

In [192]:
grand_concat = pd.concat([fchap, fcompte, fref])
filtre_lib_court = grand_concat[grand_concat.duplicated(subset = ['Nomenclature', 'Code', 'Lib_court'], keep = False)].sort_values(['Nomenclature', 'Code']).shape[0]
ignore_lib_court = grand_concat[grand_concat.duplicated(subset = ['Nomenclature', 'Code'], keep = False)].sort_values(['Nomenclature', 'Code']).shape[0]

print('Ensemble des fonctions compte, chapitre et ref possèdant le même code, nomenclature et lib', filtre_lib_court)
print('Ensemble des fonctions compte, chap et ref possédant le même code et nomenclature', ignore_lib_court)

Ensemble des fonctions compte, chapitre et ref possèdant le même code, nomenclature et lib 324
Ensemble des fonctions compte, chap et ref possédant le même code et nomenclature 386


On a des codes similaire par nomenclature, mais le libelle n'est pas le même et le reste varie (DROIS etc)
On va fouiller sur les lignes budget pour voir si on peut trouver des restrictions 

In [124]:
df_table_budget['ContFon'].value_counts()

ContFon
020    403
01     255
928    233
923    222
938    199
      ... 
512      1
812      1
810      1
002      1
04       1
Name: count, Length: 109, dtype: int64

In [129]:
df_9_compte[~df_9_compte['ContFon'].isna()][['Id_Fichier', 'Nomenclature',
        'Fonction', 'ContFon', 'Lib_court', 'REquip', 'ROES', 'ROIS', 'RR', 'RegrTotalise', 'Supprime']]

Id_Fichier         Nomenclature Fonction ContFon  \
1766     776798  M14-M14_COM_SUP3500     9230     923   
1767     776798  M14-M14_COM_SUP3500    92311     923   
1768     776798  M14-M14_COM_SUP3500    90026     900   
1769     776798  M14-M14_COM_SUP3500    92322     923   
1770     776798  M14-M14_COM_SUP3500    92020     920   
...         ...                  ...      ...     ...   
9460     806326  M14-M14_COM_SUP3500       95      95   
9461     806326  M14-M14_COM_SUP3500       95      95   
9462     806326  M14-M14_COM_SUP3500       95      95   
9463     806326  M14-M14_COM_SUP3500       95      95   
9464     806326  M14-M14_COM_SUP3500       95      95   

                                    Lib_court REquip ROES ROIS   RR  \
1766                         Services communs      0            923   
1767   Expression musicale, lyrique et choré.      0            923   
1768            Cimetières et pompes funèbres      1            900   
1769                                   Musées      0            923   
1770     Administration générale collectivité      0            920   
...                                       ...    ...  ...  ...  ...   
9460  Produits des cessions d'immobilisations      0             95   
9461  Produits des cessions d'immobilisations      0             95   
9462  Produits des cessions d'immobilisations      0             95   
9463  Produits des cessions d'immobilisations      0             95   
9464  Produits des cessions d'immobilisations      0             95   

     RegrTotalise Supprime  
1766         None     None  
1767         None     None  
1768         None     None  
1769         None     None  
1770         None     None  
...           ...      ...  
9460         None     None  
9461         None     None  
9462         None     None  
9463         None     None  
9464         None     None  

[2630 rows x 11 columns]

In [151]:
df_9_compte[df_9_compte['Lib_court'].isna()].drop_duplicates(['Nomenclature','Fonction'])

id Id_Fichier         Nomenclature  Exer_x TypOpBudg Operation  \
45    348583     784192  M14-M14_COM_SUP3500    2021      None      None   
580   364723     791323  M14-M14_COM_SUP3500    2021      None      None   
722   371092     794536  M14-M14_COM_SUP3500    2021      None      None   
854   378041     795520              M52-M52    2021      None      None   
7070  504228     820964            M57-M57_D    2022      None      None   
7323  516270     789641              M52-M52    2021      None      None   
9477  579065     732817  M14-M14_COM_SUP3500    2021      None      None   
9607  582011     651459              M57-M57    2020      None      None   

      Nature ContNat              LibCpte Fonction ContFon ArtSpe    CodRD  \
45     60612     011                 None       94    None   None  Dépense   
580    74741    None                 None        9    None  False  Recette   
722     1326      13  B - PG3 - BATIMENTS       96    None   None  Recette   
854   204141     204                 None      928    None  False  Dépense   
7070   60613     011                 None       90      90  False  Dépense   
7323  204121     204                 None       96    None  False  Dépense   
9477   20422     204                 None       97    None   None  Dépense   
9607   64138    None                 None        9    None  False  Dépense   

      MtBudgPrec  MtRARPrec  MtPropNouv    MtPrev                     OpBudg  \
45        2313.0        NaN         NaN       0.0  Opération réelle ou mixte   
580       5000.0        NaN         NaN       0.0  Opération réelle ou mixte   
722      27886.8        NaN         NaN       0.0  Opération réelle ou mixte   
854          NaN        NaN         NaN       0.0  Opération réelle ou mixte   
7070         NaN        NaN         NaN       0.0  Opération réelle ou mixte   
7323   2300000.0        0.0         0.0       0.0  Opération réelle ou mixte   
9477         NaN        NaN    955000.0  955000.0  Opération réelle ou mixte   
9607         0.0        0.0         0.0       0.0  Opération réelle ou mixte   

      CredOuv   MtReal  MtRAR3112 ContOp OpeCpteTiers  \
45        NaN      NaN        NaN   None         None   
580       NaN      NaN        NaN   None         None   
722       NaN      NaN        NaN   None         None   
854   59000.0      NaN        NaN   None         None   
7070      NaN    157.5        NaN   None         None   
7323      0.0      0.0  2300000.0   None         None   
9477      NaN      NaN        NaN   None         None   
9607      0.0  15393.0        0.0   None         None   

                                                  MtSup  MtSup_APVote  \
45          {"@Code": "BudgetHorsRAR", "@V": "2313.00"}           NaN   
580         {"@V": "5000.00", "@Code": "BudgetHorsRAR"}           NaN   
722   [{"@Code": "BudgetHorsRAR", "@V": "27886.80"},...           NaN   
854                                                None           NaN   
7070                                               None           NaN   
7323  [{"@Code": "BudgetHorsRAR", "@V": "0"}, {"@Cod...           NaN   
9477                                               None           NaN   
9607                                               None           NaN   

      MtSup_Brut  MtSup_BudgetHorsRAR MtSup_Comp  MtSup_ICNE  MtSup_ICNEPrec  \
45           NaN               2313.0       None         NaN             NaN   
580          NaN               5000.0       None         NaN             NaN   
722          NaN              27886.8       None         NaN             NaN   
854          NaN                  NaN       None         NaN             NaN   
7070         NaN                  NaN       None         NaN             NaN   
7323         NaN                  0.0       None         NaN             NaN   
9477         NaN                  NaN       None         NaN             NaN   
9607         NaN                  NaN       None         NaN             NaN   

      Mt

In [156]:
fref[fref['Code'] == '928']

Exer Nomenclature Code Lib_court Libelle
107  2020      M52-M52  928    Autres  Autres

Pour fonction, on doit prendre que fonction_compte et fonction_referentiel 

In [166]:
df_contfon = df_table_budget[~df_table_budget['ContFon'].isna()]
df_contfon_unique = df_contfon.drop_duplicates(subset = ['Nomenclature', 'ContFon'])
df_contfon_unique[df_contfon_unique['ContFon'] == '02']

id Id_Fichier          Nomenclature  Exer TypOpBudg Operation  \
212654  559332     629635  M14-M14_CCAS_SUP3500  2019      None      None   

       Nature ContNat LibCpte Fonction ContFon ArtSpe    CodRD  MtBudgPrec  \
212654    205    None    None       02      02  False  Dépense         NaN   

        MtRARPrec  MtPropNouv  MtPrev                     OpBudg  CredOuv  \
212654        NaN         NaN     0.0  Opération réelle ou mixte      NaN   

         MtReal  MtRAR3112 ContOp OpeCpteTiers MtSup  MtSup_APVote  \
212654  1326.48        NaN   None         None  None           NaN   

        MtSup_Brut  MtSup_BudgetHorsRAR MtSup_Comp  MtSup_ICNE  \
212654         NaN                  NaN       None         NaN   

        MtSup_ICNEPrec  MtSup_MtOpeCumul  MtSup_MtOpeInfo  MtSup_Net  \
212654             NaN               NaN              NaN        NaN   

       MtSup_ProdChaRat MtSup_RARPrec                         CaracSup  \
212654             None          None  {"@Code": "Section", "@V": "I"}   

        CaracSup_TypOpe CaracSup_Section CaracSup_ChapSpe  \
212654              NaN                I             None   

       CaracSup_ProgAutoLib CaracSup_ProgAutoNum CaracSup_VirCredNum  \
212654                 None                 None                None   

        CaracSup_CodeRegion  
212654                  NaN

In [170]:
fref[fref['Code'] == '02']

Exer          Nomenclature Code                            Lib_court  \
2  2020  M14-M14_CCAS_SUP3500   02  Administration générale du C.C.A.S.   
2  2020               M52-M52   02              Administration générale   
2  2020               M57-M57   02              Administration générale   
2  2020               M71-M71   02              Administration générale   
2  2020   M14-M14_COM_SUP3500   02              Administration générale   

                               Libelle  
2  Administration générale du C.C.A.S.  
2              Administration générale  
2              Administration générale  
2              Administration générale  
2              Administration générale

resumé hypothèses : 

Fonction ==> Compte et ref
ContFon ==> Chap et ref 

Compte contient les codes de Chap mais n'a pas les précisions (None partout)

In [185]:
fcomptechap = pd.concat([fcompte, fchap])
fduplic = fcomptechap[fcomptechap.duplicated(['Nomenclature', 'Code', 'Lib_court'], keep = False)].sort_values(['Nomenclature', 'Code'])
#fduplic[(~fduplic['DEquip'].isna()) | (~fduplic['Section'].isna())]
fduplic

Exer         Nomenclature Code DEquip  DOES  DOIS    DR  \
1    2020  M14-M14_COM_SUP3500  900   None  None  None  None   
18   2020  M14-M14_COM_SUP3500  900   None  None  None  None   
14   2020  M14-M14_COM_SUP3500  901   None  None  None  None   
19   2020  M14-M14_COM_SUP3500  901   None  None  None  None   
22   2020  M14-M14_COM_SUP3500  902   None  None  None  None   
..    ...                  ...  ...    ...   ...   ...   ...   
18   2020              M71-M71  952    NaN   NaN   NaN   NaN   
244  2020              M71-M71  953      0   953               
19   2020              M71-M71  953    NaN   NaN   NaN   NaN   
245  2020              M71-M71  954      0                     
38   2020              M71-M71  954    NaN   NaN   NaN   NaN   

                                    Lib_court  \
1    Services généraux administrat° publiques   
18   Services généraux administrat° publiques   
14            Sécurité et salubrité publiques   
19            Sécurité et salubrité publiques   
22                   Enseignement - Formation   
..                                        ...   
18                         Dépenses imprévues   
244    Virement à la section d'investissement   
19     Virement à la section d'investissement   
245   Produits des cessions d'immobilisations   
38    Produits des cessions d'immobilisations   

                                               Libelle REquip  ROES  ROIS  \
1    Services généraux des administrations publique...   None  None  None   
18   Services généraux des administrations publique...   None  None  None   
14                     Sécurité et salubrité publiques   None  None  None   
19                     Sécurité et salubrité publiques   None  None  None   
22                            Enseignement - Formation   None  None  None   
..                                                 ...    ...   ...   ...   
18                                  Dépenses imprévues    NaN   NaN   NaN   
244             Virement à la section d'investissement      0               
19              Virement à la section d'investissement    NaN   NaN   NaN   
245            Produits des cessions d'immobilisations      0               
38             Produits des cessions d'immobilisations    NaN   NaN   NaN   

       RR RegrTotalise Supprime SupprimeDepuis Section Special TypeChapitre  
1    None         None     None           None     NaN     NaN          NaN  
18   None         None     None           None       I       0            C  
14   None         None     None           None     NaN     NaN          NaN  
19   None         None     None           None       I       0            C  
22   None         None     None           None     NaN     NaN          NaN  
..    ...          ...      ...            ...     ...     ...          ...  
18    NaN          NaN      NaN            NaN       F       0            L  
244               None     None           None     NaN     NaN          NaN  
19    NaN          NaN      NaN            NaN       F       0            L  
245   954         None     None           None     NaN     NaN          NaN  
38    NaN          NaN      NaN            NaN       I       0            L  

[324 rows x 19 columns]

Hypothèse 2 semi foireuse : 

Il y a des cas où en effet la ligne compte n'aura pas d'autres informations que le libelle ex : M14-M14_COM_SUP3500 code 900 

Mais il y a des cas où compte possède des informations, ex : M71-M71 code 953

In [187]:
fcomptechap[(fcomptechap['Nomenclature'] == 'M71-M71') & (fcomptechap['Code'] == '953')]

Exer Nomenclature Code DEquip DOES DOIS   DR  \
244  2020      M71-M71  953      0  953             
19   2020      M71-M71  953    NaN  NaN  NaN  NaN   

                                  Lib_court  \
244  Virement à la section d'investissement   
19   Virement à la section d'investissement   

                                    Libelle REquip ROES ROIS   RR  \
244  Virement à la section d'investissement      0                  
19   Virement à la section d'investissement    NaN  NaN  NaN  NaN   

    RegrTotalise Supprime SupprimeDepuis Section Special TypeChapitre  
244         None     None           None     NaN     NaN          NaN  
19           NaN      NaN            NaN       F       0            L

In [188]:
fcomptechap[(fcomptechap['Nomenclature'] == 'M14-M14_COM_SUP3500') & (fcomptechap['Code'] == '900')]

Exer         Nomenclature Code DEquip  DOES  DOIS    DR  \
1   2020  M14-M14_COM_SUP3500  900   None  None  None  None   
18  2020  M14-M14_COM_SUP3500  900    NaN   NaN   NaN   NaN   

                                   Lib_court  \
1   Services généraux administrat° publiques   
18  Services généraux administrat° publiques   

                                              Libelle REquip  ROES  ROIS  \
1   Services généraux des administrations publique...   None  None  None   
18  Services généraux des administrations publique...    NaN   NaN   NaN   

      RR RegrTotalise Supprime SupprimeDepuis Section Special TypeChapitre  
1   None         None     None           None     NaN     NaN          NaN  
18   NaN          NaN      NaN            NaN       I       0            C

In [222]:
fcompte['DOES'].value_counts()

DOES
       1075
926       3
946       3
953       3
914       1
934       1
939       1
Name: count, dtype: int64

In [3]:
ncompte = pd.read_parquet('./parquet/Nature_Compte_2020.parquet')

In [7]:
ncompte[(ncompte['Code'] == "2051") & (ncompte['Nomenclature'] == 'M14-M14_COM_500_3500')]

Code DEquip DOES DOIS  DR                       Lib_court  \
275  2051      1  040  041  20  Concessions, droits similaires   

                              Libelle REquip ROES ROIS  RR RegrTotalise  \
275  Concessions et droits similaires      1  040  041  20         None   

    Supprime SupprimeDepuis  Exer          Nomenclature PourEtatSeul  
275     None           None  2020  M14-M14_COM_500_3500         None

In [16]:
df_table_budget[ 
        (df_table_budget['Exer'] == 2020) & 
        (df_table_budget['Fonction'] == '2121') &
        (df_table_budget['LibCpte'].str.contains('MATERIEL SCOLAIRE'))]\
        [['Id_Fichier', 'Operation', 'TypOpBudg', 'Nature', 'ContNat', 'LibCpte', 'Fonction']]

Id_Fichier Operation TypOpBudg Nature ContNat            LibCpte  \
1198406     690975      9019      None   2051    9019  MATERIEL SCOLAIRE   
1198408     690975      9019      None   2183    9019  MATERIEL SCOLAIRE   
1198411     690975      9019      None   2184    9019  MATERIEL SCOLAIRE   
1198413     690975      9019      None   2188    9019  MATERIEL SCOLAIRE   

        Fonction  
1198406     2121  
1198408     2121  
1198411     2121  
1198413     2121

In [18]:
df_table_budget[(df_table_budget['Id_Fichier'] == '690975') & (df_table_budget['Fonction'] == '2121')].shape[0]

7